File: src/main/C++/reverseKElements.cpp
/*
Reverse First k Elements of Queue
*/

#include<queue>
#include<iostream>
#include<stack>
using namespace std;

void showQ(queue<int> q){
    while(!q.empty())
    {
        cout<<q.front()<<' ';
        q.pop();
    }
    cout<<endl;
}

queue<int> reverseKElements(queue<int> q,int k){
    queue<int> rev_q;
    stack<int> tmp_s;
    int i = 0;
    
    while(!q.empty())
    {
        if(i<k){
            tmp_s.push(q.front());
            q.pop();

            if(i==k-1){
                while(!tmp_s.empty()){
                    rev_q.push(tmp_s.top());
                    tmp_s.pop();
                }
            }

        }
        else{
            rev_q.push(q.front());
            q.pop();
        }


        i++;
    }

    return rev_q;
}

int main(){
    queue<int> q;
    int k = 9;
    while(k--)
        q.push(9-k);
    showQ(q);
    showQ(reverseKElements(q,5));

}

File: src/main/python/reverseKElements.py
'''
Reverse First k Elements of Queue
'''


def reverseKelements(data:list,k:int)->list:
    rev_q = list()
    stack = list()
    for i,e in enumerate(data):
        if i<k:
            stack.append(e)
            if i==k-1:
                for j in range(len(stack)):
                    rev_q.append(stack[len(stack)-1-j])
        else:
            rev_q.append(e)

    if k>len(data):
        for j in range(len(stack)):
            rev_q.append(stack[len(stack)-1-j])


    return rev_q


if __name__ == "__main__":
    print(reverseKelements([1,2,3,4,5,6,7,8,9],0))
    print(reverseKelements([1,2,3,4,5,6,7,8,9],1))
    print(reverseKelements([1,2,3,4,5,6,7,8,9],5))
    print(reverseKelements([1,2,3,4,5,6,7,8,9],9))

    print(reverseKelements([1,2,3,4,5,6,7,8,9],10))
    print(reverseKelements([],10))